src/Pos/ExecMode.hs

```diff
@@ -0,0 +1,149 @@
+@
+
+The issue with creating a newtype per mode is that we don't have
+the expressive power to implement a class/type instance for all modes at
+once. For example, one would need to derive 'Functor', 'MonadIO', 'MonadMask',
+and other classes, for every newtype. Type families and ambiguous type variables
+make the issue worse because deriving can't handle them, so 'MonadBaseControl'
+and 'Ether.MonadReader' would need manually defined instances, and 'Mockable'-related
+type families would need numerous instances as well.
+
+What we want is to implement some class and type instances for all modes at
+once. The 'ExecMode' newtype makes it possible.
+
+Instead of
+
+@
+    newtype PatakMode ssc =
+        PatakMode (ReaderT (PatakCtx ssc) Production)
+@
+
+we write
+
+@
+    data PATAK ssc
+    type PatakMode ssc = ExecMode (PATAK ssc)
+    type instance ExecModeM (PATAK ssc) = ReaderT (PatakCtx ssc) Production
+@
+
+This way we can talk about some @'ExecMode' mode@, i.e. abstract over the
+@mode@ parameter.
+
+-}
+module Pos.ExecMode
+    ( ExecMode(..)
+    , ExecModeM
+    , _ExecMode
+    , module Pos.ExecMode.Context
+    ) where
+
+import           Universum
+
+import           Control.Lens                (makePrisms)
+import           Control.Monad.Base          (MonadBase)
+import           Control.Monad.Fix           (MonadFix)
+import qualified Control.Monad.Reader        as Mtl
+import           Control.Monad.Trans.Control (MonadBaseControl (..))
+import           Data.Coerce                 (coerce)
+import qualified Ether
+import           Ether.Internal              (HasLens (..))
+import           Mockable                    (ChannelT, Counter, Distribution, Gauge,
+                                              MFunctor' (..), Mockable (..), Promise,
+                                              SharedAtomicT, SharedExclusiveT, ThreadId)
+import           System.Wlog                 (CanLog)
+
+import           Pos.Util.Util               (PowerLift (..))
+
+-- Re-exports.
+import           Pos.ExecMode.Context
+
+-- | The inner workings of an execution mode.
+-- Example: @type instance ExecModeM Buba = ReaderT BubaEnv IO@.
+type family ExecModeM mode :: * -> *
+
+-- | A newtype over an execution mode. The mode itself is a parameter and
+-- can be abstracted over. Keep the @mode@ parameter small, without structure
+-- (ideally it's a single type name), to avoid the issues that plagued nested
+-- transformer stacks.
+newtype ExecMode mode a = ExecMode { unExecMode :: ExecModeM mode a }
+
+makePrisms ''ExecMode
+
+deriving instance Functor (ExecModeM mode) => Functor (ExecMode mode)
+deriving instance Applicative (ExecModeM mode) => Applicative (ExecMode mode)
+deriving instance Monad (ExecModeM mode) => Monad (ExecMode mode)
+deriving instance MonadIO (ExecModeM mode) => MonadIO (ExecMode mode)
+deriving instance (Monad b, MonadBase b (ExecModeM mode)) => MonadBase b (ExecMode mode)
```

Comment:
It should, but GHC doesn't see it for some reason — simply removing the `Monad b` constraint here doesn't work (I tried).